### PR TITLE
Improve performance of Char#to_s

### PR DESCRIPTION
**Pull request by BlobCodes**
I already posted this on 24th July in #7393 as an alternative implementation
I basically just glued String.new together with String#each_byte
Replaces #7393

Improvement: ~200% (One-byte characters)

<details>
<summary>Benchmark code</summary>

```crystal
require "benchmark"

puts "Benchmark 5: Char#to_s (99f8cd6326a98f40c49b32e39cd8a0cb3359508e)"

struct Char
  def new_to_s : String
    c = ord
    if c < 0x80
      # 0xxxxxxx
      String.new(1) do |pointer|
        pointer[0] = c.to_u8
        {1, 1}
      end
    elsif c <= 0x7ff
      # 110xxxxx  10xxxxxx
      String.new(2) do |pointer|
        pointer[0] = (0xc0 | c >> 6).to_u8
        pointer[1] = (0x80 | c & 0x3f).to_u8
        {2, 1}
      end
    elsif c <= 0xffff
      # 1110xxxx  10xxxxxx  10xxxxxx
      String.new(3) do |pointer|
        pointer[0] = (0xe0 | (c >> 12)).to_u8
        pointer[1] = (0x80 | ((c >> 6) & 0x3f)).to_u8
        pointer[2] = (0x80 | (c & 0x3f)).to_u8
        {3, 1}
      end
    else
      # 11110xxx  10xxxxxx  10xxxxxx  10xxxxxx
      String.new(4) do |pointer|
        pointer[0] = (0xf0 | (c >> 18)).to_u8
        pointer[1] = (0x80 | ((c >> 12) & 0x3f)).to_u8
        pointer[2] = (0x80 | ((c >> 6) & 0x3f)).to_u8
        pointer[3] = (0x80 | (c & 0x3f)).to_u8
        {4, 1}
      end
    end
  end
end

char = 'a'
puts char

Benchmark.ips do |x|
  x.report("old") { char.to_s }
  x.report("new") { char.new_to_s }
end

char = 'μ'
puts char

Benchmark.ips do |x|
  x.report("old") { char.to_s }
  x.report("new") { char.new_to_s }
end

char = '😀'
puts char

Benchmark.ips do |x|
  x.report("old") { char.to_s }
  x.report("new") { char.new_to_s }
end
```
</details>

### CODE REVIEW

**Comment by straight-shoota** [comment]
Did you paste the wrong benchmark code?

**Comment by BlobCodes** [comment]
I swapped the benchmarks with #11592, now those are the correct ones.

**Comment by straight-shoota** [comment]
It seems you can get basically the same performance improvement by passing `bytesize` to `String.new` instead of hard-coded `4`.
And I think that's to be expected. Unrolling `#each_byte` manually should make no relevant semantic difference.

Could you please also add benchmarks for longer UTF-8 encodings. I don't expect to see much difference there, but it's important to have a comprehensive understanding of the effects.

**Comment by BlobCodes** [comment]
> It seems you can get basically the same performance improvement by passing bytesize to String.new instead of hard-coded 4.

Passing the bytesize to String.new is exactly as fast for 1-3 byte characters, but 7% slower for 4-byte characters.

**Comment by oprypin** [comment]
Soo that solution seems what we should prefer, then? Almost exactly as fast but so simple.

**Comment by straight-shoota** [comment]
> exactly as fast for 1-3 byte characters, but 7% slower for 4-byte characters.

I think that can be acceptable.

**Comment by BlobCodes** [comment]
> > exactly as fast for 1-3 byte characters, but 7% slower for 4-byte characters.
> 
> I think that can be acceptable.

I just attached a benchmark to my previous post.
The performance loss seems negligable (and the results fluctuate a lot, but the current implementation is fastest most of the time).

Should I use that implementation instead?

**Comment by oprypin** [comment]
> Should I use that implementation instead?

Yes, please.

File: src/char.cr [hunk 1]
```diff
@@ -873,10 +873,14 @@ struct Char
...
   # ```
   def to_s : String
-    String.new(4) do |buffer|
-      appender = buffer.appender
-      each_byte { |byte| appender << byte }
-      {appender.size, 1}
+    bytesize = self.bytesize
+    String.new(bytesize) do |pointer|
+      i = 0
+      self.each_byte do |byte|
```
Comment (Sija) [inline src/char.cr:879]: No need for `self` (and trailing whitespace 😅 )

```suggestion
      each_byte do |byte|
```

Comment (BlobCodes) [reply src/char.cr]: I honestly have no idea where that whitespace came from, thanks!

**Comment by straight-shoota** [comment]
Why are you unwrapping the pointer appender? I wouldn't expect that to make a difference in performance.

**Review by straight-shoota — APPROVED** [review]

**Review by oprypin — APPROVED** [review]

**Comment by BlobCodes** [comment]
I just noticed that this implementation has now basically become equivalent to @r00ster91's original implementation (b690fb53404d50b4d371c6d6226b7a376b1a4364).

Because of that, I also benchmarked this against [asterite's implementation](https://github.com/crystal-lang/crystal/pull/7393#issuecomment-461575678), but for me there's no performance difference.

Also, whenever I try one of these benchmarks the results fluctuate a ton, I'm starting to hate micro-benchmarking.

**Comment by asterite** [comment]
There's also the question of whether turning a `Char` into a `String` is a common operation that warrants micro-optimizing.

**Comment by oprypin** [comment]
It's faster and it looks good. I don't know what else we'd want.

**Comment by straight-shoota** [comment]
Good point @asterite 
But using the bytesize for allocation looks like a reasonable improvement, both in theory and in practice. It's a tiny change with a relatively high impact.

**Pull request merged**